test/parallel/test-async-wrap-GH13045.js

```diff
@@ -0,0 +1,61 @@
+'use strict';
+const common = require('../common');
+
+// Refs: https://github.com/nodejs/node/issues/13045
+// An HTTP Agent reuses a TLSSocket, and makes a failed call to `asyncReset`.
+
+const assert = require('assert');
+const https = require('https');
+const fs = require('fs');
+
+const serverOptions = {
+  key: fs.readFileSync(`${common.fixturesDir}/keys/agent1-key.pem`),
+  cert: fs.readFileSync(`${common.fixturesDir}/keys/agent1-cert.pem`),
+  ca: fs.readFileSync(`${common.fixturesDir}/keys/ca1-cert.pem`)
+};
+
+const server = https.createServer(serverOptions, (req, res) => {
+  res.end('hello world\n');
+});
+
+server.listen(0, common.localhostIPv4, common.mustCall(function() {
+  const port = this.address().port;
+  const clientOptions = {
+    agent: new https.Agent({
+      keepAlive: true,
+      maxSockets: 50,
+      rejectUnauthorized: false,
+    }),
+    hostname: common.localhostIPv4,
+    port: port,
+    path: '/',
+    method: 'GET'
+  };
+
+  const req = https.request(clientOptions, common.mustCall((res) => {
+    assert.strictEqual(res.statusCode, 200);
+    res.on('error', (err) => assert.fail(err));
+    res.socket.on('error', (err) => assert.fail(err));
+    res.resume();
+    // drain the socket and wait for it to be free to reuse
+    res.socket.once('free', () => {
+      // This is the pain point. Internally the Agent will call
+      // `socket._handle.asyncReset()` and if the _handle does not implement
+      // `asyncReset` this will throw TypeError
+      const req2 = https.request(clientOptions);
+      req2.on('response', common.mustCall((res2) => {
+        assert.strictEqual(res.statusCode, 200);
+
+        res2.on('error', (err) => assert.fail(err));
+        res2.socket.on('error', (err) => assert.fail(err));
+        res2.destroy();
+        server.close();
+      }));
+      req2.end();
+    });
+  }));
+
+  req.on('error', (err) => assert.fail(err));
```

Comment:
See https://github.com/nodejs/node/pull/13092#discussion_r117214885 and https://github.com/nodejs/node/pull/13092#discussion_r117258971